tachyon/crypto/commitments/fri/two_adic_fri_pcs.h

```diff
@@ -0,0 +1,380 @@
+                                  base::bits::StrictLog2(batch_max_height);
+            size_t reduced_index = index >> bits_reduced;
+            const std::vector<std::vector<F>>& opened_values =
+                input_proof[round].opened_values;
+
+            CHECK(mmcs_.VerifyOpeningProof(commits_by_round[round], batch_dims,
+                                           reduced_index, opened_values,
+                                           input_proof[round].opening_proof));
+
+            for (size_t batch_idx = 0; batch_idx < vals_size; ++batch_idx) {
+              const Coset& mat_domain = domains_by_round[round][batch_idx];
+              const std::vector<std::tuple<ExtField, std::vector<ExtField>>>&
+                  mat_points_and_values = claim[batch_idx];
+              size_t log_height =
+                  mat_domain.domain()->log_size_of_group() + fri_.log_blowup;
+              size_t bits_reduced = log_global_max_height - log_height;
+              size_t rev_reduced_index =
+                  base::bits::ReverseBitsLen(index >> bits_reduced, log_height);
+              F w;
+              CHECK(F::GetRootOfUnity(1 << log_height, &w));
+              F x = F::FromMontgomery(F::Config::kSubgroupGenerator) *
+                    w.Pow(rev_reduced_index);
+
+              if (reduced_openings.find(log_height) == reduced_openings.end()) {
+                reduced_openings.try_emplace(
+                    log_height,
+                    std::make_tuple(ExtField::One(), ExtField::Zero()));
+              }
+              for (size_t i = 0; i < mat_points_and_values.size(); ++i) {
+                const ExtField& z = std::get<0>(mat_points_and_values[i]);
+                const std::vector<ExtField>& ps_at_z =
+                    std::get<1>(mat_points_and_values[i]);
+                CHECK_EQ(ps_at_z.size(), opened_values[i].size());
+                for (size_t j = 0; j < ps_at_z.size(); ++j) {
+                  const ExtField& p_at_z = ps_at_z[j];
+                  ExtField quotient =
+                      unwrap((-p_at_z + ExtField(opened_values[batch_idx][j])) /
+                             (-z + ExtField(x)));
+                  std::tuple<ExtField, ExtField>& reduced_opening =
+                      reduced_openings[log_height];
+                  std::get<1>(reduced_opening) +=
+                      std::get<0>(reduced_opening) * quotient;
+                  std::get<0>(reduced_opening) *= alpha;
+                }
+              }
+            }
+          }
+          int idx = 0;
+          std::vector<std::tuple<size_t, ExtField>> ret(
+              reduced_openings.size());
+          for (auto it = reduced_openings.rbegin();
+               it != reduced_openings.rend(); ++it) {
+            ret[idx++] = std::make_tuple(it->first, std::get<1>(it->second));
+          }
+          return ret;
+        });
+  }
+
+ private:
+  std::map<ExtField, std::vector<ExtField>> ComputeInverseDenominators(
```

Comment:
Please use `absl::flat_hash_map` See https://abseil.io/docs/cpp/guides/container and https://github.com/Plonky3/Plonky3/blob/main/util/src/linear_map.rs